### PR TITLE
feat(prospective-parachains): implements `GetBackableCandidates`

### PR DESCRIPTION
**Pull request by DanielDDHM**
## Changes

implements GetBackableCandidates

## Tests

Unit Tests: Core logic is tested in isolation using testify for assertions

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Issues

- Closes https://github.com/ChainSafe/gossamer/issues/4310

### CODE REVIEW

**Comment by P1sar** [comment]
Hey @DanielDDHM! Thank you for your contribution! Please can you assign corresponding issue to your self, this will helps us track progress. Thank you! 
If you have any questions, please do not hesitate to ask

**Comment by DanielDDHM** [comment]
> Hey @DanielDDHM! Thank you for your contribution! Please can you assign corresponding issue to your self, this will helps us track progress. Thank you! If you have any questions, please do not hesitate to ask

Hello @P1sar I will be assign and focus on this task again asap, at the moment I am focusing on performing this one https://github.com/ChainSafe/gossamer/issues/4390

anything else, i will let you know, thanks.

**Comment by DanielDDHM** [comment]
@haikoschol @EclesioMeloJunior @P1sar a small review please.

File: dot/parachain/prospective-parachains/prospective-parachains.go [hunk 1]
```diff
@@ -80,3 +91,77 @@ func (*ProspectiveParachains) ProcessBlockFinalizedSignal(parachaintypes.BlockFi
 	// NOTE: this subsystem does not process block finalized signal
 	return nil
 }
+
+func (pp *ProspectiveParachains) getBackableCandidates(
+	msg GetBackableCandidates,
+) {
```
Comment (EclesioMeloJunior) [inline dot/parachain/prospective-parachains/prospective-parachains.go:97]:
```suggestion
	msg GetBackableCandidates) {
```

File: dot/parachain/prospective-parachains/prospective-parachains.go [hunk 2]
```diff
@@ -80,3 +91,77 @@ func (*ProspectiveParachains) ProcessBlockFinalizedSignal(parachaintypes.BlockFi
...
+			relayParentHash, paraId,
+		)
+		responseChan <- []parachaintypes.CandidateHashAndRelayParent{}
+		return
+	}
+
+	// Retrieve backable candidates from the fragment chain
+	backableCandidates := chain.FindBackableChain(ancestors, requestedQty)
+	if len(backableCandidates) == 0 {
+		logger.Tracef(
```
Comment (haikoschol) [inline dot/parachain/prospective-parachains/prospective-parachains.go]: Why is this one `trace` and the others `debug`?

File: dot/parachain/prospective-parachains/prospective-parachains.go [hunk 3]
```diff
@@ -6,12 +6,23 @@ import (
...
+	View                *View
+}
+
+type View struct {
+	ActiveLeaves   map[common.Hash]bool
+	PerRelayParent map[common.Hash]*RelayParentData
+}
+
+type RelayParentData struct {
+	FragmentChains map[parachaintypes.ParaID]*fragmentChain
```
Comment (haikoschol) [inline dot/parachain/prospective-parachains/prospective-parachains.go]: These types are only used inside this package, so don't need to be exported, right?

```suggestion
type view struct {
	activeLeaves   map[common.Hash]bool
	perRelayParent map[common.Hash]*relayParentData
}

type relayParentData struct {
	fragmentChains map[parachaintypes.ParaID]*fragmentChain
```

File: dot/parachain/prospective-parachains/prospective_parachains_test.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		{
+			CandidateHash: parachaintypes.CandidateHash{
+				Value: common.Hash{0x11},
+			},
+			CandidateRelayParent: common.Hash{0x21},
+		},
+	}
+}
+
+func TestGetBackableCandidates(t *testing.T) {
```
Comment (haikoschol) [inline dot/parachain/prospective-parachains/prospective_parachains_test.go]: Please add more tests to cover the various cases that result in an empty response.

You probably want to use a generated mock for that. [Here](https://github.com/ChainSafe/gossamer/blob/feat/parachain/dot/parachain/backing/mocks_generate_test.go) is an example for how to do that.

File: dot/parachain/prospective-parachains/prospective-parachains.go [hunk 5]
```diff
@@ -57,7 +68,7 @@ func (pp *ProspectiveParachains) processMessage(msg any) {
 	case CandidateBacked:
 		panic("not implemented yet: see issue #4309")
 	case GetBackableCandidates:
-		panic("not implemented yet: see issue #4310")
+		go pp.getBackableCandidates(msg)
```
Comment (haikoschol) [inline dot/parachain/prospective-parachains/prospective-parachains.go]: If we run this in a goroutine, access to `pp.View` needs to be thread-safe. Did the design suggest to do this? The method shouldn't take very long to execute. 🤔

**Review by haikoschol — CHANGES_REQUESTED** [review]

File: dot/parachain/prospective-parachains/prospective_parachains_test.go (hunk 4 above)
Comment (DanielDDHM) [reply dot/parachain/prospective-parachains/prospective_parachains_test.go]: @haikoschol i trying to refactor these tests and i will let you know

File: dot/parachain/prospective-parachains/prospective-parachains.go (hunk 5 above)
Comment (EclesioMeloJunior) [reply dot/parachain/prospective-parachains/prospective-parachains.go]: we should not spawn a goroutine here, we should only call `pp.getBackableCandidates(msg)` as a normal function call

File: dot/parachain/prospective-parachains/prospective_parachains_test.go [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+		MinRelayParentNumber: 5,
+	}
+
+	mockScope, err := newScopeWithAncestors(mockRelayParent, baseConstraints, nil, 10, ancestors)
+	assert.NoError(t, err)
+
+	parentHash1, err := parentHead1.Hash()
+	if err != nil {
+		t.Fatalf("failed to hash parentHead2: %v", err)
+	}
```
Comment (EclesioMeloJunior) [inline dot/parachain/prospective-parachains/prospective_parachains_test.go]:
```suggestion
	assert.NoError(t, err)
```

**Review by EclesioMeloJunior — APPROVED** [review]

**Review by haikoschol — CHANGES_REQUESTED** [review]
@DanielDDHM `TestGetBackableCandidates/candidates_found` is failing in CI and when I run it locally.

**Comment by DanielDDHM** [comment]
> @DanielDDHM `TestGetBackableCandidates/candidates_found` is failing in CI and when I run it locally.

@haikoschol could you take a look again?

**Review by haikoschol — APPROVED** [review]
